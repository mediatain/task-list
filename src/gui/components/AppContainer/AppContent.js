import styled from 'styled-components';

export default styled.main`
  padding: 3rem 1rem;
  overflow-x: hidden;
`;
