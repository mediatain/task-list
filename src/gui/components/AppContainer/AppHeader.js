import styled from 'styled-components';
import { headerColor, textColorLight } from '../../colors';
import { elevation4 } from '../../elevations';

export default styled.header`
  background-color: ${headerColor};
  font-size: 1.2rem;
  font-weight: 600;
  line-height: 3.5rem;
  color: ${textColorLight};
  padding-left: 1rem;
  box-shadow: ${elevation4};
`;
