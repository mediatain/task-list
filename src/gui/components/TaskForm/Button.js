import styled, { css } from 'styled-components';
import { taskItemColor, textColorDark } from '../../colors';
import { elevation2, elevation8 } from '../../elevations';

export default styled.button.attrs({
  type: 'button',
})`
  background: ${taskItemColor};
  color: ${textColorDark};
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  font-weight: 500;
  border-radius: 5px;
  text-transform: uppercase;
  border: none;
  elevation: above;
  cursor: pointer;
  white-space: nowrap;
  outline: none;
  box-shadow: ${elevation2};
  transition-property: box-shadow, opacity;
  transition: 150ms ease-in-out;
  
  &:hover {
    box-shadow: ${elevation8};
  }
  
  ${({ invisible }) => invisible && css`
    opacity: 0;
    overflow: hidden;
  `}
`;
