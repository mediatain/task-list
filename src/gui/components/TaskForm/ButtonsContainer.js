import styled from 'styled-components';

export default styled.div`
  display: flex;
  margin-left: .7rem;

  @media (max-width: 38rem) {
    flex-direction: column;
    justify-content: stretch;
  }
`;
