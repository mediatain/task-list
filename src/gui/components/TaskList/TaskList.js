import React from 'react';
import TaskItem from '../TaskItem/TaskItem';
import EmptyListMotivator from './EmptyListMotivator';
import styled from 'styled-components';

const StyledTaskList = styled.ul`
  max-width: 40rem;
  margin: 2rem auto;
  padding: 0;
`;

export default function TaskList({
  tasks = [],
  onToggleTask = () => {},
}) {
  return (
    <StyledTaskList>
      {tasks.map(task => (
        <TaskItem
          key={task.id}
          task={task}
          onCheck={() => onToggleTask(task.id)}
        />
      ))}
      {tasks.length === 0 && (
        <EmptyListMotivator>Add your first task!</EmptyListMotivator>
      )}
    </StyledTaskList>
  )
}
