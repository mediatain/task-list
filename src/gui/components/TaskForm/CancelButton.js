import styled from 'styled-components';
import Button from './Button';
import { cancelColor } from '../../colors';

export default styled(Button).attrs({
  type: 'reset',
})`
  background: ${cancelColor};

  @media (max-width: 38rem) {
    padding: .25rem 1rem;
    font-size: 1rem;
    flex-grow: 1;
  }
`;
