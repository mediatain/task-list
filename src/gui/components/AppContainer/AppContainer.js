import React from 'react';
import styled from 'styled-components';
import AppContent from './AppContent';
import AppHeader from './AppHeader';
import TaskList from '../TaskList/TaskList';
import TaskForm from '../TaskForm/TaskForm';
import { backgroundColor } from '../../colors';

const StyledAppContainer = styled.div`
  background-color: ${backgroundColor};
  min-height: 100vh;
  min-width: 20rem;
`;

export default function AppContainer({
  onAddTask,
  onToggleTask,
  tasks,
  title,
}) {
  return (
    <StyledAppContainer>
      <AppHeader>{title}</AppHeader>
      <AppContent>
        <TaskForm onAddTask={onAddTask} />
        <TaskList
          tasks={tasks}
          onToggleTask={onToggleTask}
        />
      </AppContent>
    </StyledAppContainer>
  )
}
