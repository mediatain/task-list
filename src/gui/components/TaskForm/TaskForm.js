import React, { useRef, useState } from 'react';
import FormError from './FormError';
import CancelButton from './CancelButton';
import SubmitButton from './SubmitButton';
import ButtonsContainer from './ButtonsContainer';
import Button from './Button';
import styled from 'styled-components';
import TextField from './TextField';
import InputContainer from './InputContainer';
import { cancelButtonLabel, newTaskButtonLabel, saveButtonLabel } from '../../strings';

const StyledTaskForm = styled.form`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  max-width: 40rem;
  margin: auto;
  padding: 0;
`;

export default function TaskForm({ onAddTask }) {
  const textInputRef = useRef();
  const [value, setValue] = useState('');
  const [open, setOpen] = useState(false);
  const [error, setError] = useState();

  function resetForm() {
    setError();
    setOpen(false);
    setValue('');
    if (textInputRef.current) {
      textInputRef.current.blur();
    }
  }

  function onOpen() {
    setOpen(true);
    if (textInputRef.current) {
      textInputRef.current.focus();
    }
  }

  function onSubmit(e) {
    e.preventDefault();
    try {
      onAddTask(value);
      resetForm();
    } catch (e) {
      setError(e.message);
    }
  }

  function onChangeValue(e) {
    setValue(e.target.value);
    setError(undefined);
  }

  return (
    <StyledTaskForm onSubmit={onSubmit}>
      <Button onClick={onOpen} invisible={open}>{newTaskButtonLabel}</Button>
      <InputContainer visible={open}>
        <TextField
          ref={textInputRef}
          value={value}
          onChange={onChangeValue}
        />
        <ButtonsContainer>
          <CancelButton onClick={resetForm}>{cancelButtonLabel}</CancelButton>
          <SubmitButton>{saveButtonLabel}</SubmitButton>
        </ButtonsContainer>
      </InputContainer>
      {error && <FormError message={error} />}
    </StyledTaskForm>
  );
}
