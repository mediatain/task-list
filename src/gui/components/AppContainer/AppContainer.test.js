import React from 'react';
import 'jest-styled-components';
import AppContainer from './AppContainer';
import { render } from '@testing-library/react';
import TaskForm from '../TaskForm/TaskForm';
import { shallow } from 'enzyme';
import TaskList from '../TaskList/TaskList';

jest.mock('../TaskForm/TaskForm', () => jest.fn().mockReturnValue('TaskForm'));
jest.mock('../TaskList/TaskList', () => jest.fn().mockReturnValue('TaskList'));

describe('AppContainer', () => {
  it('should match the snapshot', () => {
    const { container } = render(<AppContainer />);

    expect(container).toMatchSnapshot();
  });

  it('should pass onAddTask to TaskForm', () => {
    const onAddTask = jest.fn();

    const appContainer = shallow(<AppContainer onAddTask={onAddTask} />);

    const taskForm = appContainer.find(TaskForm);
    expect(taskForm.prop('onAddTask')).toBe(onAddTask);
  });

  it('should pass tasks to TaskList', () => {
    const tasks = ['Task 1', 'Task 2'];

    const appContainer = shallow(<AppContainer tasks={tasks} />);

    const taskForm = appContainer.find(TaskList);
    expect(taskForm.prop('tasks')).toBe(tasks);
  });

  it('should pass onToggleTask to TaskList', () => {
    const onToggleTask = jest.fn();

    const appContainer = shallow(<AppContainer onToggleTask={onToggleTask} />);

    const taskForm = appContainer.find(TaskList);
    expect(taskForm.prop('onToggleTask')).toBe(onToggleTask);
  });
});

