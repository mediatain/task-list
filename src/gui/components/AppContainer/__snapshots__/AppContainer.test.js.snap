// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer should match the snapshot 1`] = `
.c2 {
  padding: 3rem 1rem;
  overflow-x: hidden;
}

.c1 {
  background-color: #039be5;
  font-size: 1.2rem;
  font-weight: 600;
  line-height: 3.5rem;
  color: #FFFFFF;
  padding-left: 1rem;
  box-shadow: 0 4px 5px 0 rgba(0,0,0,0.14),0 1px 10px 0 rgba(0,0,0,0.12),0 2px 4px -1px rgba(0,0,0,0.20);
}

.c0 {
  background-color: #F5F5F6;
  min-height: 100vh;
  min-width: 20rem;
}

<div>
  <div
    class="c0"
  >
    <header
      class="c1"
    />
    <main
      class="c2"
    >
      TaskForm
      TaskList
    </main>
  </div>
</div>
`;
